## Notebook Structure

1. Research Objective
2. Import Libraries
3. Download Historical Data
4. Inspect the Dataset
5. Clean the Data
6. Calculate Returns
7. Analyse Return Distributions
8. Calculate Historical Volatility
9. Visualise Results
10. Initial Findings

# Market Data Analysis

## 1. Research Objective

This notebook investigates historical financial market data with the aim of understanding price behaviour, return distributions, volatility, and potential distributions between market variables. 

The analysis forms the firsts stage of a broader quantitative finance research project exploring market microstructure, derivatives pricing, and systematic tradings strategies. 

## 2. Import Libraries

In [6]:
import yfinance as yf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## 3. Download Historical Data

In [5]:
spy = yf.download("SPY", start="2015-01-01", end="2026-01-01")

spy.head()

[*********************100%***********************]  1 of 1 completed


Price,Close,High,Low,Open,Volume
Ticker,SPY,SPY,SPY,SPY,SPY
Date,,,,,
2015-01-02,169.687836,170.885565,168.655319,170.472558,121465900
2015-01-05,166.623322,168.812251,166.317701,168.647051,169632600
2015-01-06,165.053955,167.449388,164.260977,166.928996,209151400
2015-01-07,167.110687,167.449356,165.929495,166.375536,125346700
2015-01-08,170.076019,170.290791,168.498345,168.514856,147217800


The dataset contains daily OHLCV (Open, High, Low, Close, Volume) observations for SPY from January 2015 to January 2026. This provides a foundation for analysing price behaviour, returns, volatility, and market dynamics.

# 4. Inspect the Dataset

In [7]:
spy.info()

<class 'pandas.DataFrame'>
DatetimeIndex: 2766 entries, 2015-01-02 to 2025-12-31
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   (Close, SPY)   2766 non-null   float64
 1   (High, SPY)    2766 non-null   float64
 2   (Low, SPY)     2766 non-null   float64
 3   (Open, SPY)    2766 non-null   float64
 4   (Volume, SPY)  2766 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 129.7 KB


The dataset contains daily observations of SPY from January 2015 to January 2026. The dataset includes OHLCV (Open, High, Low, Close, Volume) variables, which represent daily price movements and trading activity.

The data types and number of observations were inspected to ensure the dataset is suitable for further time series analysis.

In [8]:
spy.describe()

Price,Close,High,Low,Open,Volume
Ticker,SPY,SPY,SPY,SPY,SPY
count,2766.000000,2766.000000,2766.000000,2766.000000,2.766000e+03
mean,336.144656,337.905969,334.099639,336.088144,8.608106e+07
std,139.986239,140.632974,139.202705,139.970673,4.392519e+07
min,154.161621,155.207018,152.073465,153.723227,2.027000e+07
25%,218.351181,218.755158,216.927374,218.034715,5.874842e+07
50%,299.129868,300.942459,297.430154,298.721206,7.570780e+07
75%,420.883301,423.041907,418.771148,421.211618,9.941975e+07
max,686.730530,688.003728,685.626408,686.989161,5.072443e+08


Summary statistics provide an initial overview of the distribution of SPY prices and trading volume. These statistics will later be used to investigate price behaviour, volatility, and return characteristics.

In [10]:
spy.shape # Number of rows and columns

(2766, 5)

Now we see there are 2766 daily observations of SPY covering the period from January 2025 to December 2026 with 5 OHLCV variables: Open, High, Low, CLose, and Volume.

In [11]:
spy.index

DatetimeIndex(['2015-01-02', '2015-01-05', '2015-01-06', '2015-01-07',
               '2015-01-08', '2015-01-09', '2015-01-12', '2015-01-13',
               '2015-01-14', '2015-01-15',
               ...
               '2025-12-17', '2025-12-18', '2025-12-19', '2025-12-22',
               '2025-12-23', '2025-12-24', '2025-12-26', '2025-12-29',
               '2025-12-30', '2025-12-31'],
              dtype='datetime64[s]', name='Date', length=2766, freq=None)

We can now begin to analyse and clean the data.

# 5. Clean the Data

In [13]:
spy.isnull().sum() # Identifying any missing values

Price   Ticker
Close   SPY       0
High    SPY       0
Low     SPY       0
Open    SPY       0
Volume  SPY       0
dtype: int64

In [14]:
spy.index.duplicated().sum() # Identifying any duplicates in the data

np.int64(0)

The dataset has been checked for missing values and duplicate observations, so now we can begin our analysis.

# 6. Calculate Returns

We can calculate daily percentage returns:

$$
R_t = \frac{P_t-P_{t-1}}{P_{t-1}}
$$

In [27]:
spy["Simple_Return"] = spy["Close"].pct_change() 

# Introducing a new column to compute daily percentage returns

spy.head()

Price,Close,High,Low,Open,Volume,Simple_Return,Log_Return
Ticker,SPY,SPY,SPY,SPY,SPY,,
Date,,,,,,,
2015-01-02,169.687836,170.885565,168.655319,170.472558,121465900,NaN,NaN
2015-01-05,166.623322,168.812251,166.317701,168.647051,169632600,-0.018060,-0.018225
2015-01-06,165.053955,167.449388,164.260977,166.928996,209151400,-0.009419,-0.009463
2015-01-07,167.110687,167.449356,165.929495,166.375536,125346700,0.012461,0.012384
2015-01-08,170.076019,170.290791,168.498345,168.514856,147217800,0.017745,0.017589


The daily logarithmic return is calculated as:

$$
r_t = \ln\left(\frac{P_t}{P_{t-1}}\right)
$$

where:

- $r_t$ is the logarithmic return at time $t$
- $P_t$ is the closing price at time $t$
- $P_{t-1}$ is the closing price on the previous trading day

In [22]:
spy["Log_Return"] = np.log(spy["Close"] / spy["Close"].shift(1))

# Introducing a new column to compute daily logarithmic return

spy.head()

Price,Close,High,Low,Open,Volume,Simple_Return,Log_Return
Ticker,SPY,SPY,SPY,SPY,SPY,,
Date,,,,,,,
2015-01-02,169.687836,170.885565,168.655319,170.472558,121465900,NaN,NaN
2015-01-05,166.623322,168.812251,166.317701,168.647051,169632600,-0.018060,-0.018225
2015-01-06,165.053955,167.449388,164.260977,166.928996,209151400,-0.009419,-0.009463
2015-01-07,167.110687,167.449356,165.929495,166.375536,125346700,0.012461,0.012384
2015-01-08,170.076019,170.290791,168.498345,168.514856,147217800,0.017745,0.017589


This causes some issues because the first entry of both new columns is 'NaN' which we can sort. Let us find the missing values and delete them.

In [28]:
spy.isnull().sum()

Price          Ticker
Close          SPY       0
High           SPY       0
Low            SPY       0
Open           SPY       0
Volume         SPY       0
Simple_Return            1
Log_Return               1
dtype: int64

In [30]:
spy = spy.dropna()

spy.head()

Price,Close,High,Low,Open,Volume,Simple_Return,Log_Return
Ticker,SPY,SPY,SPY,SPY,SPY,,
Date,,,,,,,
2015-01-05,166.623322,168.812251,166.317701,168.647051,169632600,-0.018060,-0.018225
2015-01-06,165.053955,167.449388,164.260977,166.928996,209151400,-0.009419,-0.009463
2015-01-07,167.110687,167.449356,165.929495,166.375536,125346700,0.012461,0.012384
2015-01-08,170.076019,170.290791,168.498345,168.514856,147217800,0.017745,0.017589
2015-01-09,168.713150,170.505597,168.101896,170.489073,158567300,-0.008013,-0.008046


# 7. Analyse Return Distributions